lightning/src/ln/channelmanager.rs

```diff
@@ -1437,121 +1447,139 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+				// as-is (and were originally 0s).
+				// Of course reverse path calculation is still pretty easy given naive routing
+				// algorithms, but this fixes the most-obvious case.
+				let mut next_bytes = [0; 32];
+				chacha_stream.read_exact(&mut next_bytes).unwrap();
+				assert_ne!(next_bytes[..], [0; 32][..]);
+				chacha_stream.read_exact(&mut next_bytes).unwrap();
+				assert_ne!(next_bytes[..], [0; 32][..]);
+			}
 
-				if payment_data.is_none() {
-					return_err!("We require payment_secrets", 0x4000|0x2000|3, &[0;0]);
-				}
+			// OUR PAYMENT!
+			// final_expiry_too_soon
+			// We have to have some headroom to broadcast on chain if we have the preimage, so make sure we have at least
+			// HTLC_FAIL_BACK_BUFFER blocks to go.
+			// Also, ensure that, in the case of an unknown payment hash, our payment logic has enough time to fail the HTLC backward
+			// before our onchain logic triggers a channel closure (see HTLC_FAIL_BACK_BUFFER rational).
+			if (msg.cltv_expiry as u64) <= self.best_block.read().unwrap().height() as u64 + HTLC_FAIL_BACK_BUFFER as u64 + 1 {
+				return_err!("The final CLTV expiry is too soon to handle", 17, &[0;0]);
+			}
+			// final_incorrect_htlc_amount
+			if next_hop_data.amt_to_forward > msg.amount_msat {
+				return_err!("Upstream node sent less than we were supposed to receive in payment", 19, &byte_utils::be64_to_array(msg.amount_msat));
+			}
+			// final_incorrect_cltv_expiry
+			if next_hop_data.outgoing_cltv_value != msg.cltv_expiry {
+				return_err!("Upstream node set CLTV to the wrong value", 18, &byte_utils::be32_to_array(msg.cltv_expiry));
+			}
 
-				// Note that we could obviously respond immediately with an update_fulfill_htlc
-				// message, however that would leak that we are the recipient of this payment, so
-				// instead we stay symmetric with the forwarding case, only responding (after a
-				// delay) once they've send us a commitment_signed!
+			let routing = match next_hop_data.format {
+				msgs::OnionHopDataFormat::Legacy { .. } => return_err!("We require payment_secrets", 0x4000|0x2000|3, &[0;0]),
+				msgs::OnionHopDataFormat::NonFinalNode { .. } => return_err!("Got non final data with an HMAC of 0", 0x4000 | 22, &[0;0]),
+				msgs::OnionHopDataFormat::FinalNode { payment_data, keysend_preimage } => {
+					if payment_data.is_some() && keysend_preimage.is_some() {
+						return_err!("We don't support MPP keysend payments", 0x4000|22, &[0;0]);
+					} else if let Some(data) = payment_data {
+						PendingHTLCRouting::Receive {
+							payment_data: data,
+							incoming_cltv_expiry: msg.cltv_expiry,
+						}
+					} else if let Some(payment_preimage) = keysend_preimage {
+						// We need to check whether the keysend preimage matches the payment hash here to
+						// prevent probing attacks wherein an attacker can use the payment processing time of a
+						// bogus keysend to determine whether we had a pending inbound payment with the same
```

Comment:
"We need to check that the sender knows the keysend preimage before to process further this payment. Otherwise an intermediary routing hop forwarding HTLC A could discover the final destination of this payment, by probing the adjacent nodes with a same-hash A keysend and observe processing discrepancies due to hash collision with the pending payment A."